### PR TITLE
Make `toDelegate` safe for function pointers

### PR DESCRIPTION
**Pull request by 0xEAB**
Revamp of #8769.

While I’m not sure whether it covers all use-cases that @Bolpat’s patch had considered and that exist (also see Dennis’ unanswered question over there), it at least passes the testsuite on my machine.

As this is a `@trusted` hack, I’d appreciate my patch to be thoroughly reviewed.

### CODE REVIEW

File: std/functional.d [hunk 1]
```diff
@@ -1814,13 +1814,42 @@ if (isCallable!(F))
     {
         return fp;
     }
+    else static if (is(F Func == Func*) && is(Func == function) && is(Func Params == __parameters))
+    {
+        return () @trusted
+        {
+            return buildDelegate(fp);
```
Comment (thewilsonator) [inline std/functional.d:1821]: So, creating a delegate out of a function pointer I get is safe,

File: std/functional.d [hunk 2]
```diff
@@ -1814,13 +1814,42 @@ if (isCallable!(F))
...
+        }();
+    }
     else static if (is(typeof(&F.opCall) == delegate)
                 || (is(typeof(&F.opCall) V : V*) && is(V == function)))
     {
         return toDelegate(&fp.opCall);
     }
     else
     {
+        return buildDelegate(fp);
```
Comment (thewilsonator) [inline std/functional.d:1831]: ... what type of objects hit this branch and why is that not a safe (i.e. trustable) thing to do?

**Review by LightBender — APPROVED** [review]

File: std/functional.d [hunk 3]
```diff
@@ -1814,13 +1814,42 @@ if (isCallable!(F))
     {
         return fp;
     }
+    else static if (is(F Func == Func*) && is(Func == function) && is(Func Params == __parameters))
+    {
+        return () @trusted
```
Comment (rikkimax) [inline std/functional.d:1819]: No check if the input function is ``@safe``, therefore this will need another branch to handle those two cases.

**Review by rikkimax — CHANGES_REQUESTED** [review]

File: std/functional.d (hunk 3 above)
Comment (0xEAB) [reply std/functional.d:1819]: I’m not sure whether I can your concern here. But I’ve added a unittests that ensures that a `function() @system` won’t get transformed into `delegate() @safe`.

**Review by rikkimax — APPROVED** [review]

File: std/functional.d (hunk 2 above)
Comment (0xEAB) [reply std/functional.d]: I’ve replaced this branch with a `static assert(false);` – as I couldn’t think of a case that would both match the parameter and satisfy the template constraint. What do you think about that?

File: std/functional.d [hunk 4]
```diff
@@ -1854,21 +1882,22 @@ if (isCallable!(F))
...
     }
 
-    uint myNum = 0;
-    auto incMyNumDel = toDelegate(&inc);
-    auto returnVal = incMyNumDel(myNum);
-    assert(myNum == 1);
+    uint myNum = 0x1337;
+    struct S1 { int opCall() { inc(myNum); return myNum; } }
+    static assert(!is(typeof(&s1.opCall) == delegate));
+    S1 s1;
```
Comment (Bolpat) [inline std/functional.d:1895]: What’s that doing?
In `static assert(!is(typeof(&s1.opCall) == delegate));`, the local variable `s1` isn’t in scope.
If you swap the lines, the `static assert` passes.

Comment (0xEAB) [reply std/functional.d:1895]: good catch

Comment (0xEAB) [reply std/functional.d:1895]: Apparently, it’s a copy&paste of this code: <https://github.com/dlang/phobos/blob/8c6fca9ccdb30b69b627808c698526a76f7c4248/std/functional.d#L1891-L1901>
<https://github.com/dlang/phobos/blame/8c6fca9ccdb30b69b627808c698526a76f7c4248/std/functional.d#L1891-L1901>

**Pull request merged**